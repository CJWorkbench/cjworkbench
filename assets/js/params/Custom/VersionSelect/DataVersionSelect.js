import React from 'react'
import PropTypes from 'prop-types'
import DataVersionModal from '../../../WorkflowEditor/DataVersionModal'
import { connect } from 'react-redux'
import { t, Trans } from '@lingui/macro'
import { withI18n } from '@lingui/react'

export class DataVersionSelect extends React.PureComponent {
  static propTypes = {
    i18n: PropTypes.shape({
      // i18n object injected by LinguiJS withI18n()
      _: PropTypes.func.isRequired
    }),
    wfModuleId: PropTypes.number.isRequired,
    currentVersionIndex: PropTypes.number, // or null for no selected version
    nVersions: PropTypes.number.isRequired, // may be 0
    isReadOnly: PropTypes.bool.isRequired
  }

  state = {
    isDataVersionModalOpen: false
  }

  handleClickOpenModal = () => this.setState({ isDataVersionModalOpen: true })
  handleCloseModal = () => this.setState({ isDataVersionModalOpen: false })

  render () {
    const { wfModuleId, currentVersionIndex, nVersions, isReadOnly, i18n } = this.props
    const { isDataVersionModalOpen } = this.state

    let inner

    if (nVersions === 0) {
      inner = (
        <>
          <div className='label'>
            <Trans id='js.params.Custom.VersionSelect.DataVersionSelect.noVersions.label'>
                Version
            </Trans>
          </div>
          <div className='no-versions'>â€“</div>
        </>
      )
    } else if (isReadOnly) {
      inner = (
        <div className='read-only'>
          <Trans id='js.params.Custom.VersionSelect.DataVersionSelect.readOnly.label' description='The parameter {0} will be the current version and {nVersions} will be the number of versions'>
            Version {nVersions - currentVersionIndex} of {nVersions}
          </Trans>
        </div>
      )
    } else {
      inner = (
        <>
          <div className='label'>
            <Trans id='js.params.Custom.VersionSelect.DataVersionSelect.selectVersion.label'>
                Version
            </Trans>
          </div>
          <button
            type='button'
            title={i18n._(t('js.params.Custom.VersionSelect.DataVersionSelect.selectVersion.hoverText')`Select version`)}
            onClick={this.handleClickOpenModal}
          >
            <Trans id='js.params.Custom.VersionSelect.DataVersionSelect.versionCount' description='The parameter {0} will be the current version and {nVersions} will be the number of versions'>
              {nVersions - currentVersionIndex} of {nVersions}
            </Trans>
          </button>
          {isDataVersionModalOpen ? (
            <DataVersionModal
              wfModuleId={wfModuleId}
              onClose={this.handleCloseModal}
            />
          ) : null}
        </>
      )
    }

    return (
      <div className='version-item'>
        {inner}
      </div>
    )
  }
}

function mapStateToProps (state, { wfModuleId }) {
  const isReadOnly = state.workflow.read_only

  const wfModule = state.wfModules[String(wfModuleId)]
  if (!wfModule || !wfModule.versions || !wfModule.versions.selected) {
    return {
      currentVersionIndex: null,
      nVersions: 0,
      isReadOnly
    }
  }

  const { versions, selected } = wfModule.versions
  const index = versions.findIndex(arr => arr[0] === selected) || null

  return {
    currentVersionIndex: index,
    nVersions: versions.length,
    isReadOnly
  }
}

export default connect(
  mapStateToProps
)(withI18n()(DataVersionSelect))
