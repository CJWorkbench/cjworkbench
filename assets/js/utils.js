// ---- Utilities ---
import * as Cookies from 'js-cookie'
import { fromByteArray as base64Encode } from 'base64-js'
import { t } from '@lingui/macro'

export function goToUrl (url) {
  window.location.href = url
}

// Current CSRF token
export const csrfToken = Cookies.get('csrftoken')

// Gets the letter coordinate of a column from its index within the column names array
export function idxToLetter (idx) {
  var letters = ''
  var cidx = parseInt(idx)
  cidx += 1
  do {
    cidx -= 1
    letters = String.fromCharCode(cidx % 26 + 65) + letters
    cidx = Math.floor(cidx / 26)
  } while (cidx > 0)
  return letters
}

// Log to Intercom, if installed
export function logUserEvent (name, metadata) {
  if (!window.APP_ID) return

  // If we're in a lesson, drop the event.
  if (window.initState && window.initState.lessonData) return

  window.Intercom('trackEvent', name, metadata)
}

export function timeDifference (start, end, i18n) {
  const ms = new Date(end) - new Date(start)
  const seconds = Math.floor(ms / 1000)
  const minutes = Math.floor(seconds / 60)
  const hours = Math.floor(minutes / 60)
  const days = Math.floor(hours / 24)
  const years = Math.floor(days / 365.25)

  if (years > 0) {
    return i18n._(t('js.util.timeDifference.ago.years')`${years}y ago`)
  } else if (days > 0) {
    return i18n._(t('js.util.timeDifference.ago.days')`${days}d ago`)
  } else if (hours > 0) {
    return i18n._(t('js.util.timeDifference.ago.hours')`${hours}h ago`)
  } else if (minutes > 0) {
    return i18n._(t('js.util.timeDifference.ago.minutes')`${minutes}m ago`)
  } else if (seconds > 0) {
    return i18n._(t('js.util.timeDifference.ago.seconds')`${seconds}s ago`)
  } else {
    return i18n._(t('js.util.timeDifference.now')`just now`)
  }
}

export function escapeHtml (str) {
  str
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&#039;')

  return str
}

/**
 * Scroll `containerEl` vertically by the smallest amount possible to put
 * `el` in view.
 *
 * Set the container CSS `scroll-behavior: smooth` to make this transition
 * look nice.
 *
 * @param el Element we want to focus
 * @param containerEl Element we'll set scrollTop on so it contains el
 * @param marginTop Minimum number of pixels between containerEl.top and el.top
 * @param marginBottom Minimum number of pixels between containerEl.bottom and el.bottom
 */
export function scrollTo (el, containerEl, marginTop, marginBottom) {
  if (marginTop === undefined) marginTop = 10
  if (marginBottom === undefined) marginBottom = 10

  const elRect = el.getBoundingClientRect()
  const containerRect = containerEl.getBoundingClientRect()

  // Calculate dy: how much do we need to add to containerEl.scrollTop to
  // make el go where we want it to go?
  let dy = 0 // Prefer not to scroll
  if (elRect.bottom + marginBottom > containerRect.bottom) {
    // Scroll down if we need to.
    // We do this before up because some els might be taller than containerEl,
    // and in that case we want scroll-up to override scroll-down (because the
    // top of el is more important than the bottom).
    dy += (elRect.bottom + marginBottom - containerRect.bottom)
  }
  if (dy + elRect.top - marginTop < containerRect.top) {
    // Scroll up if we need to.
    dy -= (containerRect.top - (elRect.top - marginTop))
  }

  containerEl.scrollTop += dy
}

/**
 * Build a unique slug with the given prefix.
 *
 * The slug uses randomly-generated bytes: 9 bytes, base64-encoded with
 * '+' and '/' changed to '-' and '_' (to fit a Django "slug" field).
 *
 * Django 'slug' regex: r'^[-a-zA-Z0-9_]+\Z'
 *
 * According to https://en.wikipedia.org/wiki/Birthday_attack, 9 bytes -- 72
 * bits -- means if clients call this 97,000 times, they have a <10^-12 chance
 * of generating a collision within a given Workflow. This should protect even
 * the most click-happy user. (6 bytes would allow 380 before the chance grows
 * higher, and 380 steps is possible.)
 *
 * (Why 9 bytes and not 8? Because base64-encoding gives the same string
 * length. 9 bytes encoded as base64 gives 12 bytes.)
 */
export function generateSlug (prefix) {
  const bytes = new Uint8Array(9)
  window.crypto.getRandomValues(bytes)
  return prefix + (
    base64Encode(bytes)
      .replace(/\+/g, '-')
      .replace(new RegExp('/', 'g'), '_')
  )
}
