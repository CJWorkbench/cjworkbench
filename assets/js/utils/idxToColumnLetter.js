/**
 * Return 'A' for 0, 'B' for 1, 'AA' for 26, and so on.
 */
export default function idxToColumnLetter (idx) {
  let letters = ''
  let cidx = parseInt(idx)
  cidx += 1
  do {
    cidx -= 1
    letters = String.fromCharCode((cidx % 26) + 65) + letters
    cidx = Math.floor(cidx / 26)
  } while (cidx > 0)
  return letters
}
