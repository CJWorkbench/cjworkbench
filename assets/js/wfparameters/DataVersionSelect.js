import React from 'react'
import { Modal, ModalHeader, ModalBody, ModalFooter } from 'reactstrap'
import dateFormat from 'dateformat'
import PropTypes from 'prop-types'


export default class DataVersionSelect extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      modalOpen: false,
      dropdownOpen: false,
      versions: {versions: [], selected: ''},
      originalSelected: null
    };
    this.toggleModal = this.toggleModal.bind(this);
    this.toggleDropdown = this.toggleDropdown.bind(this);
    this.setSelected = this.setSelected.bind(this);
    this.changeVersions = this.changeVersions.bind(this);
  }

  // Takes a date string, interpreted as UTC, and produce a string for user display in user tz
  formatDate(datestr) {

    // dates already arrive in UTC
    var d = new Date(datestr);

    // check if we have a valid date
    if (isNaN(d.getTime())) return null;

    // return time as a user-readable string - "true" to convert to UTC for tests
    if (this.props.testing) {
      return dateFormat(d, "mmm d, yyyy - hh:MM TT", true)
    } else {
      return dateFormat(d, "mmm d, yyyy - hh:MM TT")
    }
  }

  toggleModal() {
    if (!this.props.isReadOnly) {
      this.setState(Object.assign({}, this.state, { modalOpen: !this.state.modalOpen }));
    }
  }

  toggleDropdown() {
    if (this.props.isReadOnly) {
      this.setState(Object.assign({}, this.state, { dropdownOpen: !this.state.dropdownOpen }));
    }
  }

  loadVersions() {
    this.props.api.getWfModuleVersions(this.props.wfModuleId)
      .then(json => {
        this.setState(
          Object.assign({}, this.state, {versions: json, originalSelected: json.selected})
        );
      })
  }

  // Load version list / current version when first created
  componentDidMount() {
    this.loadVersions();
  }

  // If the workflow revision changes, reload the versions in case they've changed too
  componentWillReceiveProps(nextProps) {
    if (this.props.revision != nextProps.revision) {
      this.loadVersions();
    }
  }

  setSelected(date) {
    if (!this.props.isReadOnly) {
      this.setState(
        Object.assign(
          {},
          this.state,
          {versions: {'versions': this.state.versions.versions, 'selected': date}}
        )
      );
    }
  }

  changeVersions() {
    if (this.state.versions.selected !== this.state.originalSelected) {
      this.props.api.setWfModuleVersion(this.props.wfModuleId, this.state.versions.selected)
      .then(() => {
        this.setState(
          Object.assign({}, this.state, {originalSelected: this.state.versions.selected})
        )
      })
    }

    this.toggleModal();
  }

  render() {
    var versionText;
    var modalLink;

    var totalVers = this.state.versions.versions.length;

    if (totalVers > 0) {
      var curVers = totalVers - this.state.versions.versions.indexOf(this.state.versions.selected);
      versionText = "Version " + curVers + " of " + totalVers;

      modalLink =
        <div>
          <div className='open-modal t-f-blue content-4 text-center' onClick={this.toggleModal}>
                {this.formatDate(this.state.originalSelected)}
          </div>
          <Modal isOpen={this.state.modalOpen} toggle={this.toggleModal} className='dialog-window'>
            <ModalHeader toggle={this.toggleModal} >
              <div className=''>Dataset Versions</div>
            </ModalHeader>
            <ModalBody className='dialog-body'>
              <div className='scolling-list'>
                {this.state.versions.versions.map( date => {
                  return (
                    <div
                      key={date}
                      className={
                        (date == this.state.versions.selected)
                          ? 'line-item-data-selected list-test-class'
                          : 'line-item-data list-test-class'
                      }
                      onClick={() => this.setSelected(date)}
                    >
                      <span className='content-3'>{ this.formatDate(date) }</span>
                    </div>
                  );
                })}
              </div>
            </ModalBody>
            <ModalFooter className='dialog-footer'>
              <div className='button-gray mr-3 action-button test-cancel-button' onClick={this.toggleModal}>Cancel</div>
              <div className='button-blue action-button test-ok-button' onClick={this.changeVersions}>OK</div>
            </ModalFooter>
          </Modal>
        </div>

    } else {
      versionText = "No data loaded";
      modalLink = null;
    }

    return (
      <div className='version-item'>
        <div className='t-d-gray content-3 mb-2'>{versionText}</div>
        {modalLink}
      </div>
    );
  }
}

DataVersionSelect.propTypes = {
  wfModuleId:       PropTypes.number.isRequired,
  revision:         PropTypes.number.isRequired,
  api:              PropTypes.object.isRequired,
  testing:          PropTypes.bool            // for testing only
};
