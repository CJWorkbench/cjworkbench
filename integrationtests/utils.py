# Utilities for integration tests

import subprocess
import unittest
import email.message
from typing import Optional
import re

from integrationtests.browser import Browser
from integrationtests.helpers import accounts


_url_regex = re.compile('https?://[^\\s]+')

def find_url_in_email(message: email.message.Message) -> Optional[str]:
    """Return the first URL in the given message's payload, or None."""
    body = message.get_payload()
    match = _url_regex.search(body)

    if not match: return None
    return match.group(0)


def _find_server_url():
    """Get URL using `docker port`"""
    process = subprocess.run([
        'docker',
        'port',
        'cjworkbench_integrationtest_django',
        '8000/tcp'
    ], stdout=subprocess.PIPE)
    port_str = process.stdout.decode('ascii').split(':')[1].strip()
    return f"http://localhost:{port_str}"


class WorkbenchBase(unittest.TestCase):
    serve_static = True
    account_admin = accounts.AccountAdmin()
    live_server_url = _find_server_url()

    def setUp(self):
        super().setUp()

        self.account_admin.clear_data_from_previous_tests()

        #self.current_site = Site.objects.get_current()
        #self.SocialApp1 = self.current_site.socialapp_set.create(
        #    provider="facebook",
        #    name="Facebook",
        #    client_id="1234567890",
        #    secret="0987654321",
        #)
        #self.SocialApp2 = self.current_site.socialapp_set.create(
        #    provider="google",
        #    name="Google",
        #    client_id="1234567890",
        #    secret="0987654321",
        #)

        self.browser = Browser(base_url=self.live_server_url)


    def tearDown(self):
        self.browser.quit()


# Derive from this to perform all tests logged in
class LoggedInIntegrationTest(WorkbenchBase):
    def setUp(self):
        super().setUp()

        self.user = self.account_admin.create_user('user@example.org')
        self.user_email = self.account_admin.verify_user_email(self.user)

        accounts.login(self.browser, self.user.email, self.user.email)


    def tearDown(self):
        self.account_admin.destroy_user_email(self.user_email)
        self.account_admin.destroy_user(self.user)
        self.account_admin.destroy_modules()

        super().tearDown()
