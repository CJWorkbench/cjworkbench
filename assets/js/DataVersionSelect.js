import React from 'react'
import { Button, Modal, ModalHeader, ModalBody, ModalFooter } from 'reactstrap'
import { Form, FormGroup, Label, Input } from 'reactstrap'
import { Dropdown, DropdownToggle, DropdownMenu, DropdownItem } from 'reactstrap'
import { csrfToken } from './utils'
import dateFormat from 'dateformat'
import PropTypes from 'prop-types'



export default class DataVersionSelect extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      modalOpen: false,
      dropdownOpen: false,
      versions: {versions: [], selected: ''},
      originalSelected: ''
    };

    // Allow props to specify a conversion from browser time to displayed time, so tests can run in UTC (not test machine tz)
    if (props.timezoneOffset != undefined) {
      this.state.timezoneOffset = props.timezoneOffset;
    } else {
      this.state.timezoneOffset = 0; // display in browser local time
    }

    this.toggleModal = this.toggleModal.bind(this);
    this.toggleDropdown = this.toggleDropdown.bind(this);
    this.setSelected = this.setSelected.bind(this);
    this.changeVersions = this.changeVersions.bind(this);
  }

  // Takes a date string, interpreted as UTC, and produce a string for user display in user tz
  formatDate(datestr) {

    // interpret date string as UTC always (comes from server this way)
    var d = new Date(datestr + 'UTC');

    // check if we have a valid date
    if (isNaN(d.getTime())) return null;

    // Convert to milliseconds, then add tz offset (which is in minutes)
    var nd = new Date(d.getTime() + (60000*this.state.timezoneOffset));

    // return time as a user-readable string
    return dateFormat(nd, "mmmm d yyyy - hh:MM TT")
  }

  toggleModal() {
    if (!this.props.isReadOnly) {
      this.setState(Object.assign({}, this.state, { modalOpen: !this.state.modalOpen }));
    }
  }

  toggleDropdown() {
    if (this.props.isReadOnly) {
      this.setState(Object.assign({}, this.state, { dropdownOpen: !this.state.dropdownOpen }));
    }
  }

  loadVersions() {
    this.props.api.getWfModuleVersions(this.props.wfModuleId)
      .then(json => {
        this.setState(
          Object.assign({}, this.state, {versions: json, originalSelected: json.selected})
        );
      })
  }

  // Load version list / current version when first created
  componentDidMount() {
    this.loadVersions();
  }

  // If the workflow revision changes, reload the versions in case they've changed too
  componentWillReceiveProps(nextProps) {
    if (this.props.revision != nextProps.revision) {
      this.loadVersions();
    }
  }

  setSelected(date) {
    if (!this.props.isReadOnly) {
      this.setState(
        Object.assign(
          {},
          this.state,
          {versions: {'versions': this.state.versions.versions, 'selected': date}}
        )
      );
    }
  }

  changeVersions() {
    if (this.state.versions.selected !== this.state.originalSelected) {
      this.props.api.setWfModuleVersion(this.props.wfModuleId, this.state.versions.selected)
      .then(() => {
        this.setState(
          Object.assign({}, this.state, {originalSelected: this.state.versions.selected})
        )
      })
    }

    this.toggleModal();
  }

  render() {
    // TODO: Assign conditional render if module is open/closed: see WfModule 115
    // TODO: Refactor calculated classNames outside of Return statement

    return (
      <div className='version-item d-flex justify-content-center flex-column align-items-center'>
        <div className='t-d-gray content-3 mb-4'>Current Version</div>

        <div className='open-modal t-f-blue content-3 text-center' onClick={this.toggleModal}>
            {this.state.originalSelected != '' ? this.formatDate(this.state.originalSelected) : ''}
        </div>
        <Modal isOpen={this.state.modalOpen} toggle={this.toggleModal} >
          <ModalHeader toggle={this.toggleModal} >
            <div className=''>Dataset Versions</div>
          </ModalHeader>
          <ModalBody className='dialog-body'>
            <div className='scolling-list'>
              {this.state.versions.versions.map( date => {
                return (
                  <div
                    key={date}
                    className={
                      (date == this.state.versions.selected)
                        ? 'line-item-data-selected list-test-class'
                        : 'line-item-data  list-test-class'
                    }
                    onClick={() => this.setSelected(date)}
                  >
                    <span className='content-3'>{ this.formatDate(date) }</span>
                  </div>
                );
              })}
            </div>
          </ModalBody>
          <ModalFooter className='dialog-footer'>
            <div className='button-blue action-button' onClick={this.toggleModal}>Cancel</div>
            <div className='button-blue action-button' onClick={this.changeVersions}>OK</div>
          </ModalFooter>
        </Modal>
      </div>
    );
  }
}

DataVersionSelect.propTypes = {
  wfModuleId:       PropTypes.number.isRequired,
  revision:         PropTypes.number.isRequired,
  api:              PropTypes.object.isRequired,
  timezoneOffset:   PropTypes.number
};
