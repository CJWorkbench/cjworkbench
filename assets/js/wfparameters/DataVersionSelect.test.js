import React from 'react'
import DataVersionSelect  from './DataVersionSelect'
import { mount, ReactWrapper } from 'enzyme'
import { okResponseMock, jsonResponseMock } from '../utils'

jest.useFakeTimers();

describe('DataVersionSelect', () => {

  var mockVersions = {
    versions: [
      '2017-07-10 17:57:58.324Z',
      '2017-06-10 17:57:58.324Z',
      '2017-05-10 17:57:58.324Z',
      '2017-04-10 17:57:58.324Z',
      '2017-03-10 17:57:58.324Z'
    ],
    selected: '2017-04-10 17:57:58.324Z'
  };
  var api;
  var wrapper;
  var modalLink;

  // Mount is necessary to invoke componentDidMount()
  beforeEach(() => {
    api = {
      getWfModuleVersions: jsonResponseMock(mockVersions),
      setWfModuleVersion: okResponseMock(),
    };

    wrapper = mount(
      <DataVersionSelect
        isReadOnly={false}
        wfModuleId={808}
        revision={202}
        api={api}
        testing={true}
      />
    );

    modalLink = wrapper.find('div.open-modal')
  });

  it('Renders correctly when in Private mode, and selection is confirmed when user hits OK', (done) => {

    // should call API for its data on componentDidMount
    expect(api.getWfModuleVersions.mock.calls.length).toBe(1);

    expect(wrapper).toMatchSnapshot();  // 1

    // Start with dialog closed
    expect(wrapper.state().modalOpen).toBe(false)

    // link renders with default text before versions are loaded
    expect(modalLink.text()).toEqual("-");

    // give versions a chance to load
    setImmediate( () => {
      expect(modalLink.text()).toEqual("Apr 10, 2017 - 05:57 PM");

      expect(modalLink).toHaveLength(1);
      modalLink.simulate('click');
      expect(wrapper.state().modalOpen).toBe(true);

      // Need setImmediate to give modal a chance to fill with data, API returns a promise that must resolve
      setImmediate( () => {
        expect(wrapper.state().modalOpen).toBe(true)

        // The insides of the Modal are a "portal", that is, attached to root of DOM, not a child of Wrapper
        // So find them, and make a new Wrapper
        // Reference: "https://github.com/airbnb/enzyme/issues/252"
        let modal_element = document.getElementsByClassName('dialog-window');
        expect(modal_element.length).toBe(1);
        let modal = new ReactWrapper(modal_element[0], true)

        expect(modal).toMatchSnapshot(); // 2
        expect(modal.find('.dialog-body')).toHaveLength(1);

        // check that the versions have loaded and are displayed in list
        expect(wrapper.state().versions).toEqual(mockVersions);
        let versionsList = modal.find('.list-test-class');
        expect(versionsList).toHaveLength(5);

        // check that date has correctly converted to readable string
        // ?????

        // filter list to grab first item
        let firstVersion = versionsList.filterWhere(n => n.key() == '2017-07-10 17:57:58.324Z');
        expect(firstVersion).toHaveLength(1);
        firstVersion.simulate('click');

        expect(wrapper.state().versions.selected).toEqual('2017-07-10 17:57:58.324Z');
        expect(wrapper.state().originalSelected).toEqual('2017-04-10 17:57:58.324Z');

        let okButton = modal.find('.test-ok-button');
        expect(okButton).toHaveLength(1);
        okButton.first().simulate('click');

        // state needs to update and modal needs to close
        setImmediate( () => {
          // timezone bugs
          expect(wrapper).toMatchSnapshot(); // 3
          expect(wrapper.state().modalOpen).toBe(false);
          expect(wrapper.state().originalSelected).toEqual('2017-07-10 17:57:58.324Z');
          expect(api.getWfModuleVersions.mock.calls.length).toBe(1);
          expect(api.setWfModuleVersion.mock.calls.length).toBe(1);
          done();
        });
      });
    });
  });

  // Pared-down version of first test
  it('Does not save selection when user hits Cancel', (done) => {

    expect(api.getWfModuleVersions.mock.calls.length).toBe(1);

    modalLink.simulate('click');
    expect(wrapper.state().modalOpen).toBe(true);

    setImmediate( () => {
      let modal_element = document.getElementsByClassName('dialog-window');
      // select the second element in array for this test
      expect(modal_element.length).toBe(2);
      let modal = new ReactWrapper(modal_element[1], true);

      // check that the versions have loaded and are displayed in list
      expect(wrapper.state().versions).toEqual(mockVersions);
      let versionsList = modal.find('.list-test-class');
      expect(versionsList).toHaveLength(5);
      let lastVersion = versionsList.filterWhere(n => n.key() == '2017-03-10 17:57:58.324Z');
      lastVersion.simulate('click');

      expect(wrapper.state().versions.selected).toEqual('2017-03-10 17:57:58.324Z');
      expect(wrapper.state().originalSelected).toEqual('2017-04-10 17:57:58.324Z');

      let cancelButton = modal.find('.test-cancel-button');
      cancelButton.first().simulate('click');

      setImmediate( () => {
        expect(wrapper).toMatchSnapshot();              // 4
        expect(wrapper.state().modalOpen).toBe(false);
        expect(wrapper.state().originalSelected).toEqual('2017-04-10 17:57:58.324Z');
        expect(api.setWfModuleVersion.mock.calls.length).toBe(0); // never called because user cancelled
        done();
      });
    });
  });

  it('Does not open modal when in read-only mode', (done) => {
    let readOnlywrapper = mount(<DataVersionSelect
      isReadOnly={true}
      wfModuleId={808}
      revision={202}
      api={api}
      testing={true}
    />);

    expect(api.getWfModuleVersions.mock.calls.length).toBe(2); // 2 not 1 because beforeEach mounted "wrapper" already

    let readOnlyModalLink = readOnlywrapper.find('div.open-modal')

    readOnlyModalLink.simulate('click');
    expect(readOnlywrapper.state().modalOpen).toBe(false);

    done();
  });

  it('Displays empty when no versions available', (done) => {

    var emptyApi = {
      getWfModuleVersions: jsonResponseMock({versions: [], selected: null}),
      setWfModuleVersion: okResponseMock(),
    };

    let wrapper2 = mount(<DataVersionSelect
      isReadOnly={true}
      wfModuleId={808}
      revision={202}
      api={emptyApi}
      testing={true}
    />);

    setImmediate( () => {
      expect(emptyApi.getWfModuleVersions.mock.calls.length).toBe(1);

      var modalLink2 = wrapper2.find('div.open-modal');
      expect(modalLink2.text()).toEqual("-");

      expect(wrapper2).toMatchSnapshot();
      done();
    });

  });
});
