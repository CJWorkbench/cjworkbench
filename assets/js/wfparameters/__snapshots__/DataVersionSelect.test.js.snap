// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataVersionSelect Displays empty when no versions available 1`] = `
<DataVersionSelect
  api={
    Object {
      "getWfModuleVersions": [Function],
      "setWfModuleVersion": [Function],
    }
  }
  isReadOnly={true}
  revision={202}
  testing={true}
  wfModuleId={808}
>
  <div
    className="version-item"
  >
    <div
      className="t-d-gray content-3 mb-2"
    >
      Current Version
    </div>
    <div
      className="open-modal t-f-blue content-4 text-center"
      onClick={[Function]}
    >
      -
    </div>
    <Modal
      autoFocus={true}
      backdrop={true}
      backdropTransitionTimeout={150}
      className="dialog-window"
      fade={true}
      isOpen={false}
      keyboard={true}
      modalTransitionTimeout={300}
      toggle={[Function]}
      zIndex={1050}
    />
  </div>
</DataVersionSelect>
`;

exports[`DataVersionSelect Does not save selection when user hits Cancel 1`] = `
<DataVersionSelect
  api={
    Object {
      "getWfModuleVersions": [Function],
      "setWfModuleVersion": [Function],
    }
  }
  isReadOnly={false}
  revision={202}
  testing={true}
  wfModuleId={808}
>
  <div
    className="version-item"
  >
    <div
      className="t-d-gray content-3 mb-2"
    >
      Current Version
    </div>
    <div
      className="open-modal t-f-blue content-4 text-center"
      onClick={[Function]}
    >
      Apr 10, 2017 - 05:57 PM
    </div>
    <Modal
      autoFocus={true}
      backdrop={true}
      backdropTransitionTimeout={150}
      className="dialog-window"
      fade={true}
      isOpen={false}
      keyboard={true}
      modalTransitionTimeout={300}
      toggle={[Function]}
      zIndex={1050}
    />
  </div>
</DataVersionSelect>
`;

exports[`DataVersionSelect Renders correctly when in Private mode, and selection is confirmed when user hits OK 1`] = `
<DataVersionSelect
  api={
    Object {
      "getWfModuleVersions": [Function],
      "setWfModuleVersion": [Function],
    }
  }
  isReadOnly={false}
  revision={202}
  testing={true}
  wfModuleId={808}
>
  <div
    className="version-item"
  >
    <div
      className="t-d-gray content-3 mb-2"
    >
      Current Version
    </div>
    <div
      className="open-modal t-f-blue content-4 text-center"
      onClick={[Function]}
    >
      -
    </div>
    <Modal
      autoFocus={true}
      backdrop={true}
      backdropTransitionTimeout={150}
      className="dialog-window"
      fade={true}
      isOpen={false}
      keyboard={true}
      modalTransitionTimeout={300}
      toggle={[Function]}
      zIndex={1050}
    />
  </div>
</DataVersionSelect>
`;

exports[`DataVersionSelect Renders correctly when in Private mode, and selection is confirmed when user hits OK 2`] = `
<div
  className="modal-dialog dialog-window"
  role="document"
>
  <div
    className="modal-content"
  >
    <ModalHeader
      tag="h4"
      toggle={[Function]}
      wrapTag="div"
    >
      <div
        className="modal-header"
      >
        <h4
          className="modal-title"
        >
          <div
            className=""
          >
            Dataset Versions
          </div>
        </h4>
        <button
          aria-label="Close"
          className="close"
          onClick={[Function]}
          type="button"
        >
          <span
            aria-hidden="true"
          >
            Ã—
          </span>
        </button>
      </div>
    </ModalHeader>
    <ModalBody
      className="dialog-body"
      tag="div"
    >
      <div
        className="dialog-body modal-body"
      >
        <div
          className="scolling-list"
        >
          <div
            className="line-item-data list-test-class"
            onClick={[Function]}
          >
            <span
              className="content-3"
            >
              Jul 10, 2017 - 05:57 PM
            </span>
          </div>
          <div
            className="line-item-data list-test-class"
            onClick={[Function]}
          >
            <span
              className="content-3"
            >
              Jun 10, 2017 - 05:57 PM
            </span>
          </div>
          <div
            className="line-item-data list-test-class"
            onClick={[Function]}
          >
            <span
              className="content-3"
            >
              May 10, 2017 - 05:57 PM
            </span>
          </div>
          <div
            className="line-item-data-selected list-test-class"
            onClick={[Function]}
          >
            <span
              className="content-3"
            >
              Apr 10, 2017 - 05:57 PM
            </span>
          </div>
          <div
            className="line-item-data list-test-class"
            onClick={[Function]}
          >
            <span
              className="content-3"
            >
              Mar 10, 2017 - 05:57 PM
            </span>
          </div>
        </div>
      </div>
    </ModalBody>
    <ModalFooter
      className="dialog-footer"
      tag="div"
    >
      <div
        className="dialog-footer modal-footer"
      >
        <div
          className="button-gray mr-3 action-button test-cancel-button"
          onClick={[Function]}
        >
          Cancel
        </div>
        <div
          className="button-blue action-button test-ok-button"
          onClick={[Function]}
        >
          OK
        </div>
      </div>
    </ModalFooter>
  </div>
</div>
`;

exports[`DataVersionSelect Renders correctly when in Private mode, and selection is confirmed when user hits OK 3`] = `
<DataVersionSelect
  api={
    Object {
      "getWfModuleVersions": [Function],
      "setWfModuleVersion": [Function],
    }
  }
  isReadOnly={false}
  revision={202}
  testing={true}
  wfModuleId={808}
>
  <div
    className="version-item"
  >
    <div
      className="t-d-gray content-3 mb-2"
    >
      Current Version
    </div>
    <div
      className="open-modal t-f-blue content-4 text-center"
      onClick={[Function]}
    >
      Jul 10, 2017 - 05:57 PM
    </div>
    <Modal
      autoFocus={true}
      backdrop={true}
      backdropTransitionTimeout={150}
      className="dialog-window"
      fade={true}
      isOpen={false}
      keyboard={true}
      modalTransitionTimeout={300}
      toggle={[Function]}
      zIndex={1050}
    />
  </div>
</DataVersionSelect>
`;
